ext/http/00_serve.js

```diff
@@ -0,0 +1,466 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+const core = globalThis.Deno.core;
+const primordials = globalThis.__bootstrap.primordials;
+
+const { BadResourcePrototype } = core;
+import { InnerBody } from "ext:deno_fetch/22_body.js";
+import { Event } from "ext:deno_web/02_event.js";
+import { toInnerResponse } from "ext:deno_fetch/23_response.js";
+import { fromInnerRequest } from "ext:deno_fetch/23_request.js";
+import { AbortController } from "ext:deno_web/03_abort_signal.js";
+import {
+  _eventLoop,
+  _idleTimeoutDuration,
+  _idleTimeoutTimeout,
+  _protocol,
+  _readyState,
+  _rid,
+  _role,
+  _server,
+  _serverHandleIdleTimeout,
+  SERVER,
+  WebSocket,
+} from "ext:deno_websocket/01_websocket.js";
+import {
+  getReadableStreamResourceBacking,
+  readableStreamForRid,
+  ReadableStreamPrototype,
+  writableStreamForRid,
+} from "ext:deno_web/06_streams.js";
+const {
+  ObjectPrototypeIsPrototypeOf,
+  SafeSet,
+  SafeSetIterator,
+  SetPrototypeAdd,
+  SetPrototypeDelete,
+  Symbol,
+  TypeError,
+  Uint8ArrayPrototype,
+} = primordials;
+
+// Symbol we use to communicate to the rest of Deno that we handle upgrades ourselves
+const _wantsUpgrade = Symbol("_wantsUpgrade");
+const _upgraded = Symbol("_upgraded");
+
+class RequestHeaders {
+  #slabId;
+
+  get(header) {
+    const value = core.ops.op_get_request_header(this.#slabId, header);
+    return value;
+  }
+}
+
+class InnerRequest {
+  #slabId;
+  #headers;
+  #context;
+  #methodAndUri;
+  #streamRid;
+  #body;
+  #upgraded;
+
+  constructor(slabId, context) {
+    this.#slabId = slabId;
+    this.#context = context;
+    this.#upgraded = false;
+  }
+
+  close() {
+    if (this.#streamRid !== undefined) {
+      core.close(this.#streamRid);
+      this.#streamRid = undefined;
+    }
+    this.#slabId = undefined;
+  }
+
+  get [_upgraded]() {
+    return this.#upgraded;
+  }
+
+  [_wantsUpgrade](upgradeType, ...originalArgs) {
+    this.#upgraded = true;
+
+    // upgradeHttp is async
+    // TODO(mmastrac)
+    if (upgradeType == "upgradeHttp") {
+      throw "upgradeHttp is unavailable in Deno.serve at this time";
+    }
+
+    // upgradeHttpRaw is async
+    // TODO(mmastrac)
+    if (upgradeType == "upgradeHttpRaw") {
+      throw "upgradeHttp is unavailable in Deno.serve at this time";
+    }
+
+    // upgradeWebSocket is sync
+    if (upgradeType == "upgradeWebSocket") {
```

Comment:
The new implementation of `upgradeWebSocket` does not seem correct. The actual upgrade should not begin until the token `Response` from the `Deno.upgradeWebSocket` function has been returned from the serve handler. `Deno.upgradeWebSocket` should not initiate I/O. It only sets up the token response that can initiate an upgrade later. (It entangles a `WebSocket` object with a token `Response` object)

Reply:
Ah yeah. This should probably be reflected in the documentation for `upgradeWebSocket`.

While I understand not wanting to break existing code, the old designs of these APIs are particularly nasty and deadlock-prone (`upgradeHttp` even has a deadlock warning in the doc page). `Deno.serve` is unstable and I think it's better to do things the right way in this API and migrate `serveHttp` over to the same in Deno 2.0. 

As `Deno.serve` is unstable, there should be no breaking API concerns. Creating a response that the user has to shuttle off to the right place is a massive footgun and it seems like a big mistake to do that here rather than update the documentation